### PR TITLE
RavenDB-13107 - Basic support for sharded queries

### PR DESCRIPTION
**Pull request by ayende**
This is still very raw and hacky, but it shows the overall structure that I think we'll have.

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs [hunk 1]
```diff
@@ -148,7 +148,7 @@ public async Task Try()
                     writer.WriteComma();
                     writer.WritePropertyName("Includes");
                     var includes = new List<Document>();
-                    includeCmd.Fill(includes);
+                    includeCmd.Fill(includes, false);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:151]: issue to handle that?

File: src/Raven.Server/Documents/Queries/GraphQueryRunner.cs [hunk 2]
```diff
@@ -182,7 +182,7 @@ public async Task WriteDetailedQueryResult(IndexQueryServerSide indexQuery, Quer
                         }
                     }
 
-                    idc.Fill(final.Includes);
+                    idc.Fill(final.Includes, false);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Queries/GraphQueryRunner.cs:185]: Also an issue?

File: src/Raven.Server/Documents/Queries/IndexQueryServerSide.cs [hunk 3]
```diff
@@ -84,6 +95,29 @@ public IndexQueryServerSide(QueryMetadata metadata)
 
         public bool IsStream;
 
+        private BlittableJsonReaderObject _asJson;
+
+        public BlittableJsonReaderObject ToJson(JsonOperationContext context)
+        {
+            if (_asJson != null) 
+                return _asJson;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Queries/IndexQueryServerSide.cs:103]: Can it move between contexts? Just wonder if `Clone` here is needed.

File: src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+            {
+                var tasks = new List<Task>();
+
+                foreach (var (shardIndex, cmd) in _commands)
+                {
+                    var task = _parent.ShardedContext.RequestExecutors[shardIndex].ExecuteAsync(cmd, cmd.Context);
+                    tasks.Add(task);
+                }
+
+                await Task.WhenAll(tasks);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs:319]: Also here we need some handling for partial failures?

File: src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+                                return -1;
+                            return AlphaNumericFieldComparator.StringAlphanumComparer.Instance.Compare(xVal, yVal);
+                        }
+                    case Queries.AST.OrderByFieldType.Random:
+                        return (_random ??= new Random()).Next(0, int.MaxValue);
+                 
+                    case Queries.AST.OrderByFieldType.Custom:
+                        throw new NotSupportedException("Custom sorting is not supported in sharding as of yet");
+                    case Queries.AST.OrderByFieldType.Score:
+                    case Queries.AST.OrderByFieldType.Distance:
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs:492]: TODO?

File: src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs [hunk 6]
```diff
@@ -0,0 +1,622 @@
...
+                            //   support including from another shard, so we'll need to do that again for missing includes
+                            //   That means also recording the include() call from JS on missing values that we'll need to rerun on
+                            //   other shards
+                            var includeTask = queryProcessor.HandleIncludes();
+                            if (includeTask.IsCompleted == false)
+                            {
+                                await includeTask.AsTask();
+                            }
+                            
+                            queryProcessor.MergeResults();
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs:105]: We can have partial errors in `ExecuteShardedOperations` **and** / **or** in `HandleIncludes`, not sure how to handle all the permutations here.
Also need to consider those scenarios with respect to the reduce result.

File: src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+                        return;
+                    }
+                }
+
+                
+                // * For paging queries, we modify the limits on the query to include all the results from all
+                //   shards if there is an offset. But if there isn't an offset, we can just get the limit from
+                //   each node and then merge them
+                if (_query.Offset > 0)
+                {
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs:240]: I think, it would be nice to see here maybe some performance hint, to suggest to avoid this and use `startWith` instead?

File: src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs [hunk 8]
```diff
@@ -0,0 +1,622 @@
...
+                    }
+                    else
+                    {
+                        queryTemplate.Modifications[nameof(IndexQuery.QueryParameters)] = modifiedArgs = new DynamicJsonValue();
+                    }
+
+                    int limit = ((_query.Limit  ?? 0) + (_query.Offset ?? 0))  * _parent.ShardedContext.NumberOfShardNodes;
+                    if (limit < 0) // overflow
+                        limit = int.MaxValue;
+                    modifiedArgs["__raven_limit"] = limit;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs:262]: not sure where `__raven_limit` is wired and how it translated to RQL `limit`..

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 9]
```diff
@@ -2134,9 +2134,20 @@ private void UpdateDatabase(ClusterOperationContext context, string type, Blitta
 
                     if (databaseRecordJson == null)
                     {
-                        if (updateCommand.ErrorOnDatabaseDoesNotExists)
-                            throw DatabaseDoesNotExistException.CreateWithMessage(databaseName, $"Could not execute update command of type '{type}'.");
-                        return;
+                        var shardIndex = valueNameLowered.AsSpan().LastIndexOf((byte)'$');
+                        if (shardIndex != -1) // try loading the sharded version, instead
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:2138]: Why? I don't think it is a valid scenario and we should handle that higher in the stack?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 10]
```diff
@@ -2377,6 +2389,8 @@ private void SqueezeDatabasesToSingleNodeCluster(ClusterOperationContext context
                 using (var rawRecord = ReadRawDatabaseRecord(context, name))
                 {
                     var topology = rawRecord.Topology;
+                    if (topology == null)
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:2392]: Is that a TODO?

File: src/Raven.Studio/wwwroot/version.txt [hunk 11]
```diff
@@ -1 +1 @@
-﻿{ "Version": "5.1.0-custom-51" }
+﻿{ "Version": "5.1.5-custom-51" }
```
Comment (karmeli87) [inline src/Raven.Studio/wwwroot/version.txt:1]: why the change?

Comment (karmeli87) [reply src/Raven.Studio/wwwroot/version.txt:1]: and we need to change to 6.0 :-)

File: src/Sparrow/Json/JsonDeserializationBase.cs [hunk 12]
```diff
@@ -375,6 +375,9 @@ private static ParameterExpression GetParameter(Type type, Dictionary<Type, Para
         private static Dictionary<string, T> ToDictionaryOfPrimitive<T>(BlittableJsonReaderObject json, string name, JsonDeserializationStringDictionaryAttribute jsonDeserializationDictionaryAttribute)
             where T : struct
         {
+            if (json == null)
```
Comment (karmeli87) [inline src/Sparrow/Json/JsonDeserializationBase.cs:378]: why we added all those `null` checks?

File: src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs:151]: https://issues.hibernatingrhinos.com/issue/RavenDB-16279

File: src/Raven.Server/Documents/Queries/GraphQueryRunner.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Documents/Queries/GraphQueryRunner.cs:185]: Not planning to support graph queries on sharded dbs at this point at all

File: src/Raven.Server/Documents/Queries/IndexQueryServerSide.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Documents/Queries/IndexQueryServerSide.cs:103]: It shouldn't, a query has a single context for its duration, unless I missed something.

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 9 above)
Comment (ayende) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:2138]: Because in this way, pretty much all our existing infrastructure behaves in the same manner. We just create "virtual" sharded dbs, but we have to specialize a lot less to support it.

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 10 above)
Comment (ayende) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:2392]: Not quite, this can happen when you are reading sharded database, since the root doesn't have a topology.

Comment (ayende) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:2392]: https://issues.hibernatingrhinos.com/issue/RavenDB-16280

File: src/Raven.Studio/wwwroot/version.txt (hunk 11 above)
Comment (ayende) [reply src/Raven.Studio/wwwroot/version.txt:1]: Accidental

File: src/Sparrow/Json/JsonDeserializationBase.cs (hunk 12 above)
Comment (ayende) [reply src/Sparrow/Json/JsonDeserializationBase.cs:378]: Because you might actually get nulls here, and we do when working with null topology, for example.
This is deserialization code and should handle this scenario.
